src/systems/equation_systems.C

```diff
@@ -855,65 +856,104 @@ EquationSystems::build_elemental_solution_vector (std::vector<Number> & soln,
-
+  // Resolve class of type input and ensure that at least one of them is null
+  libmesh_error_msg_if(type && types,
+                       "Input either 'type' or 'types' in find_variable_numbers, but not both.");
+
+  std::vector<FEType> type_filter;
+  if (type)
+    type_filter.push_back(*type);
+  else if (types)
+    type_filter = *types;
+
+  // Store a copy of the valid variable names, if any. The names vector will be be repopulated with
+  // any valid names (or all if 'is_names_empty') in the system which pass through the type filter,
+  // if any. If the variable is a vector, its name will be decomposed into its separate components
+  // in accordance with build_variable_names()
+  std::vector<std::string> name_filter = names;
+  bool is_names_empty = name_filter.empty();
   names.clear();
 
+  // initialize convenience variables
+  FEType var_type;
+  std::string name;
+
+  const std::vector<std::string> component_suffix = {"_x", "_y", "_z"};
+  unsigned int dim = _mesh.mesh_dimension();
+  libmesh_error_msg_if(dim > 3, "Invalid dim in find_variable_numbers");
+
+  // Now filter through the variables in each system and store the system index and their index
+  // within that system. This way, we know where to find their data even after we sort them
+  std::vector<std::pair<unsigned int, unsigned int>> var_nums;
+
   for (const auto & pr : _systems)
     {
       const System & system = *(pr.second);
-      const unsigned int nv_sys = system.n_vars();
 
-      for (unsigned int var=0; var < nv_sys; ++var)
+      for (unsigned int var = 0; var < system.n_vars(); ++var)
         {
-          const std::string & name = system.variable_name(var);
-          if ((type && system.variable_type(var) != *type) ||
-              (is_filter_names && std::find(filter_names.begin(), filter_names.end(), name) == filter_names.end()))
+          // apply the type filter
+          var_type = system.variable_type(var);
+          if (type_filter.size() &&
+              std::find(type_filter.begin(), type_filter.end(), var_type) == type_filter.end())
             continue;
 
-          // Otherwise, this variable should be output
+          // apply the name filter (note that all variables pass if it is empty)
+          if (FEInterface::field_type(var_type) == TYPE_VECTOR)
+            {
+              std::vector<std::string> component_names;
+              for (unsigned int comp = 0; comp < dim; ++comp)
+                {
+                  name = system.variable_name(var) + component_suffix[comp];
+                  if (is_names_empty ||
+                      (std::find(name_filter.begin(), name_filter.end(), name) != name_filter.end()))
+                    component_names.push_back(name);
+                }
+
+              if (! component_names.empty())
+                names.insert(names.end(), component_names.begin(), component_names.end());
+              else
+                continue;
+            }
+          else /*scalar variable*/
```

Comment:
Write "scalar-valued variable" here, so we don't get confused and think this branch has something to do with `SCALAR`?

No, we don't have good nomenclature here.

Reply:
Okay!